### DISCUSSION TITLE
Private actions

### DISCUSSION THREAD

**Discussion by wtuminski** [mobxjs/mobx-state-tree#1884] [Q&A] [answered]
Hi!
I was wondering if there is a way to create a private actions (like we would do with `private` keyword in a class)? 
Sometimes I want an action to be available only in a model that owns it and not in e.g. parent models.

Thank you in advance!

**Comment by elektronik2k5** [comment]
I don't think there's a way to do that. When I needed to signal that a given view or action is an implementation detail and not part of a public API, I used the convention which predates (real) private fields and TS's `private` access modifier: via an `_` as prefix of the identifier in question.
But there may be a better way, which I'm not familiar with.

**Reply by wtuminski** [reply]
Thank you for your answer! :)  I also use the prefix but I am curious if there is another way.

**Comment by EmilTholin** [comment, marked as answer]
I agree with @elektronik2k5 and use the same convention myself.

If you really don't want to expose internal actions, you can hide them in the scope similar to what you can do with [`volatile state`](https://mobx-state-tree.js.org/concepts/volatiles), granted they are synchronous.

```ts
import { types } from "mobx-state-tree";

let Store = types
  .model({
    num: types.number,
    log: types.array(types.number)
  })
  .actions((self) => {
    const _logNum = (num: number) => {
      self.log.push(num);
    };

    return {
      setNum(num: number) {
        self.num = num;
        _logNum(num);
      }
    };
  });

const store = Store.create({ num: 1, log: [] });

store.setNum(2);
console.log(store); // { num: 2, log: [2] }
// Property '_logNum' does not exist on type...
store._logNum(3);
```

**Reply by wtuminski** [reply]
Thank you!
Good idea, I haven't thought about it.

**Reply by wtuminski** [reply]
@EmilTholin 
I am not sure if the edited code is going to work. MST might throw an error that you try to modife an object outside of an action.
The previous options with flow should work (haven't checked yet), but I think we should add a following line to the private action: `      yield Promise.resolve(()=>{})` . WDYT?

**Reply by EmilTholin** [reply]
The edited code will work as long as the hidden "action" is synchronous, as it will be considered part of the action/`flow` that calls it.

The reason I edited was because I was wrong in my first answer. The hidden `flow` could give rise to the error `a mst flow must always have a parent context`.

**Reply by wtuminski** [reply]
Understood! Thank you for your reply 🙂
